src/core/ddsc/src/shm_monitor.c

```diff
@@ -80,96 +67,55 @@ dds_return_t shm_monitor_wake_and_enable(shm_monitor_t* monitor)
 
 dds_return_t shm_monitor_attach_reader(shm_monitor_t* monitor, struct dds_reader* reader) 
 {
-
-    if(iox_listener_attach_subscriber_event(monitor->m_listener, reader->m_iox_sub, SubscriberEvent_DATA_RECEIVED, shm_subscriber_callback) != ListenerResult_SUCCESS) {
-        DDS_CLOG(DDS_LC_SHM, &reader->m_rd->e.gv->logconfig, "error attaching reader\n");    
+  enum iox_ListenerResult attach_result =
+      iox_listener_attach_subscriber_event_with_context_data(monitor->m_listener,
+                                                             reader->m_iox_sub,
+                                                             SubscriberEvent_DATA_RECEIVED,
+                                                             shm_subscriber_callback,
+                                                             &reader->m_iox_sub_context);
+  if(ListenerResult_SUCCESS != attach_result) {
+    switch (attach_result) {
+      case ListenerResult_EVENT_ALREADY_ATTACHED:{
+        break;
+      }
+      case ListenerResult_LISTENER_FULL:
+      case ListenerResult_EMPTY_EVENT_CALLBACK:
+      case ListenerResult_EMPTY_INVALIDATION_CALLBACK:
+      case ListenerResult_UNDEFINED_ERROR:
+      default: {
+        DDS_CLOG(DDS_LC_SHM, &reader->m_rd->e.gv->logconfig, "error attaching reader\n");
         return DDS_RETCODE_OUT_OF_RESOURCES;
+      }
     }
-    ++monitor->m_number_of_attached_readers;
+  }
 
-    return DDS_RETCODE_OK;
+  reader->m_iox_sub_context.monitor = &reader->m_entity.m_domain->m_shm_monitor;
+  return DDS_RETCODE_OK;
 }
 
 dds_return_t shm_monitor_detach_reader(shm_monitor_t* monitor, struct dds_reader* reader) 
 {
-    iox_listener_detach_subscriber_event(monitor->m_listener, reader->m_iox_sub, SubscriberEvent_DATA_RECEIVED); 
-    --monitor->m_number_of_attached_readers;
+    // if the reader is attached
+    if (reader->m_iox_sub_context.monitor != NULL && reader->m_iox_sub_context.parent_reader != NULL) {
+        iox_listener_detach_subscriber_event(monitor->m_listener, reader->m_iox_sub, SubscriberEvent_DATA_RECEIVED);
+        reader->m_iox_sub_context.monitor = NULL;
```

Comment:
OK.
There should be no race in the callback due to setting these NULL, as it should run to completion before detaching. This also means a blocking callback will block detach, but this was also the case before (and the callback is not blocking).